Src/LiquidProjections/Statistics/WeightedProjectionSpeedCalculator.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+
+namespace LiquidProjections.Statistics
+{
+    /// <summary>
+    /// Calculates the weighted speed in transactions per second.
+    /// </summary>
+    public class WeightedProjectionSpeedCalculator
+    {
+        private readonly TimeSpan threshold;
+
+        private readonly int maxNrOfSamples;
+
+        private readonly ConcurrentQueue<long> samples = new ConcurrentQueue<long>();
+
+        public WeightedProjectionSpeedCalculator(int maxNrOfSamples, TimeSpan threshold)
+        {
+            this.maxNrOfSamples = maxNrOfSamples;
+            this.threshold = threshold;
+        }
+
+        public DateTime LastSampleTimeStampUtc { get; private set; }
+
+        public long LastCheckpoint { get; private set; }
+
+        public bool HasBaselineBeenSet { get; private set; }
+
+        public void Record(long checkpoint, DateTime timestampUtc)
+        {
+            TimeSpan interval = timestampUtc - LastSampleTimeStampUtc;
+            if (interval > threshold)
+            {
+                long delta = checkpoint - LastCheckpoint;
+
+                samples.Enqueue((long)(delta / interval.TotalSeconds));

```

Comment:
If the speed is less than 1 transaction per second, the calculator will return zero speed
If the speed is less than 10 transactions per second, the calculation error will be greater than 10%.